### PR TITLE
[Sketcher] Ability to snap slot horizontal/vertical

### PR DESCRIPTION
**Pull request by 0penBrain**
Triggered by pressing Ctrl key when creating the slot
 Extra constraint is added in case snapping is enabled

- [X]  Your pull request is confined strictly to a single module. That is, all the files changed by your pull request are either in `App`, `Base`, `Gui` or one of the `Mod` subfolders. If you need to make changes in several locations, make several pull requests and wait for the first one to be merged before submitting the next ones
- [X]  In case your pull request does more than just fixing small bugs, make sure you discussed your ideas with other developers on the FreeCAD forum
- [X]  Your branch is [rebased](https://git-scm.com/docs/git-rebase) on latest master `git pull --rebase upstream master`
- [X]  All FreeCAD unit tests are confirmed to pass by running `./bin/FreeCAD --run-test 0`
- [X]  All commit messages are [well-written](https://chris.beams.io/posts/git-commit/) ex: `Fixes typo in Draft Move command text`
- [X]  Your pull request is well written and has a good description, and its title starts with the module name, ex: `Draft: Fixed typos`
- [X]  Commit messages include `issue #<id>` or `fixes #<id>` where `<id>` is the [FreeCAD bug tracker issue number](https://freecadweb.org/wiki/tracker#GitHub_and_MantisBT) in case a particular commit solves or is related to an existing issue on the tracker. Ex: `Draft: fix typos - fixes #0004805`

### CODE REVIEW

**Comment by freecadci** [comment]
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_5200">pipeline status</a> for feature branch [PR_5200](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_5200). Pipeline [417423669](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/417423669) was triggered at [c58128c](https://github.com/FreeCAD/FreeCAD/pull/5200/commits/c58128c4f4917c5b120c409ae4d540b721bc8ed8). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

File: src/Mod/Sketcher/Gui/DrawSketchHandler.h [hunk 1]
```diff
@@ -45,7 +45,8 @@ struct AutoConstraint
     enum TargetType
     {
         VERTEX,
-        CURVE
+        CURVE,
+        VERTEX_FOR_PRIMITIVE
```
Comment (abdullahtahiriyo) [inline src/Mod/Sketcher/Gui/DrawSketchHandler.h:49]: I would consider renaming "VERTEX_FOR_PRIMITIVE" to something showing better the intent. From the comment in the code, I get that we introduce this new enum value to support seeking for a vertex but foregoing any attempt of tangency (I did not look in deep into this detection). If that is what is intended, maybe "VERTEX_NO_TANGENCY" could be what you are looking for. Maybe I am wrong, and this mode is only useful when drawing a primitive (which is a concept from Part WB, I think, I do not remember using it in the context of the sketcher). Just, give it a thought. I won't insist on it. I did not get the naming at first and this is why I comment.

File: src/Mod/Sketcher/Gui/DrawSketchHandler.cpp [hunk 2]
```diff
@@ -600,19 +604,19 @@ void DrawSketchHandler::createAutoConstraints(const std::vector<AutoConstraint>
                                      , geoId1, posId1, geoId2);
                 } break;
             case Sketcher::Horizontal: {
-                Gui::cmdAppObjectArgs(sketchgui->getObject(), "addConstraint(Sketcher.Constraint('Horizontal',%i)) ", geoId1);
+                Gui::cmdAppObjectArgs(sketchgui->getObject(), "addConstraint(Sketcher.Constraint('Horizontal',%i)) ",
+                                      geoId2 != Constraint::GeoUndef ? geoId2 : geoId1);
```
Comment (abdullahtahiriyo) [inline src/Mod/Sketcher/Gui/DrawSketchHandler.cpp:608]: I find this geoId selection behind the scenes unexpected for a user. My rationale goes like this: the function (createAutoConstraint) takes a geoId1. For a horizontal constraint on an edge there is only one geoId. A user of the function would, I think, expect that the autoconstraint is applied on geoId1. I do understand that this is not the wanted functionality in this example. geoId2, is used for the other cases as an additional geoId for constraints needing two geoIds. If after giving it a thought, you still think this is the best way, I would appreciate if you could add a minimalistic documentation to the function indicating that if the GeoId stored in the Autoconstraint is defined, it takes precedence over the geoId passed as a parameter.

**Review by abdullahtahiriyo — APPROVED** [review]
Thanks for the functionality.

I have left a couple of comments. I can live with this PR without modification. Please, give it a thought to those comments and let me know.

File: src/Mod/Sketcher/Gui/DrawSketchHandler.h (hunk 1 above)
Comment (0penBrain) [reply src/Mod/Sketcher/Gui/DrawSketchHandler.h]: VERTEX_NO_TANGENCY is fine for me. Done ;)

File: src/Mod/Sketcher/Gui/DrawSketchHandler.cpp (hunk 2 above)
Comment (0penBrain) [reply src/Mod/Sketcher/Gui/DrawSketchHandler.cpp:611]: @abdullahtahiriyo I added a comment that hopefully may help.
Honestly, I perfectly understood what geoId2 was made for. Actually you can see in the commit message that I said all this is a "hack". :)
I thought about many other "cleaner" ways, even starting some coding. But I found each to require far bigger codebase changes, and being very fat for what it was intended. Finally I found this hack was probably a good solution and should serve its purpose gently. So I pushed the PR.
But definitely don't merge it if you're not totally fine with it. ;)

**Comment by freecadci** [comment]
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_5200">pipeline status</a> for feature branch [PR_5200](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_5200). Pipeline [417927123](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/417927123) was triggered at [70b4eb7](https://github.com/FreeCAD/FreeCAD/pull/5200/commits/70b4eb7df218bbdc4825c1b31b3bdc8df9158dce). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

**Comment by luzpaz** [comment]
@0penBrain Any chance you're adding this to the [0.20 Release page](https://wiki.freecad.org/Release_notes_0.20)? Any gifs of this feature in action?

**Comment by 0penBrain** [comment]
@luzpaz I will as soon as I managed to get some spare time. Hopefully in a few hours it's done. ;)

**Comment by 0penBrain** [comment]
@luzpaz Done.

**Pull request merged**